draft-irtf-cfrg-opaque.md

```diff
@@ -1053,6 +1056,131 @@ Steps:
 7. Output (client_private_key, response.server_public_key, export_key)
 ~~~
 
+## Protocol Compilation at Login {#protocol-compilation}
+
+OPAQUE links an OPRF with an AKE protocol for login authentication. These can be executed concurrently, and the running
+AKE can use the output values of the credential retrieval process, which includes the OPRF.
+
+OPAQUE exposes the following API for authentication:
+- `ke1, blind, client_secret = ClientInit(client_identity, password, client_info)`
+- `ke2, client_info, expected_client_mac, session_key = ServerInit(server_identity, server_private_key, server_public_key, record, credential_identifier, oprf_seed)`
+- `ke3, server_info, session_key = ClientFinish(password, blind, client_secret, client_identity, server_identity, ke1, ke2)`
+- `valid = ServerFinish(expected_client_mac, ke3)`
+
+{{opaque-client}} and {{opaque-server}} specify the inner working of these functions.
+
+The `ke1`, `ke2`, and `ke3` output values are the protocol messages sent between client and server.
+
+The return value of `ServerFinalize()` indicates the client's authenticity to the server and therefore the validity of
+`client_info` and `session_key` produced by `Response()`. If it returns `false`, then the client MUST NOT be trusted and `client_info` and `session_key` MUST NOT be used.
+
+The following values are internal state values and displayed here for exhaustivity, to be carried from one stage to the other:
```

Comment:
Since these are internal values, perhaps we can hide them under a "client_state" and "server_state" structure? Otherwise, APIs might be tempted to export them for applications to use.